sklearn/decomposition/tests/test_pca.py

```diff
@@ -49,6 +57,59 @@ def test_pca(svd_solver, n_components):
     assert_allclose(np.dot(cov, precision), np.eye(X.shape[1]), atol=1e-12)
 
 
+@pytest.mark.parametrize("density", [0.01, 0.1, 0.30])
+@pytest.mark.parametrize("n_components", [1, 2, 10, SPARSE_MAX_COMPONENTS])
```

Comment:
@ogrisel, in my experience there has been platform specific difference in numerical stability. Specifically, whether the CPUs have AVX2 instruction sets has been a confusing source of discrepencies in PCA.

I'm totally up for cutting down the test time, but am concerned about making it harder to reproduce failure cases. I think it would be really valuable to be able to deterministically run all cases. Is there a good way to mark some cases as slow?

Reply:
We can run the tests for all admissible random seeds (0 to 99) via:

```
SKLEARN_TESTS_GLOBAL_RANDOM_SEED="all" pytest sklearn/some/module.py -k test_something
```

and our nightly build will try a different seed each night and open an issue in the CI in case of failure.

Reply:
Running the tests with that command locally 41/ 11200 combinations fail, always with 1 element being not within the threshold (e.g. `SKLEARN_TESTS_GLOBAL_RANDOM_SEED="all" pytest sklearn/decomposition/tests/test_pca.py -n auto -k test_pca_sparse`)

If I increase all thresholds by a factor of 10, there are 5 cases that still fail. But only by a small margin (all of these are the same random seed, 8).

In all cases, the absolute difference is less than 1e-12.

Should I just keep bumping `rtol` up, or is there something more principled I should be doing here?

Reply:
Yes. The test should never fail (all random seeds). Reduce tolerance or increase data size a bit or modify something else.